### ISSUE TITLE
Enhanced run_tests support for pbs with prereq

### ISSUE THREAD

**Issue by jasondhales** [idaholab/moose#3152]
The bison/examples/restart example has three parts:
1. RunApp.  Run the complete analysis, writing results into gold/output.e.
2. RunApp.  Run a few steps of the analysis, writing out restart files.  Has 1 as prereq.
3. Exodiff.  Read the restart files from 2, and run to the end.  Compare to the output from 1.  Has 2 as prereq.

This setup works fine on our workstations, but it hiccups on fission since each part (1, 2, and 3) runs in a separate directory.  One way around this would be to launch one pbs job for all tests that are connected through the use of prereq.

This situation can be tested by running with --pbs -i examples on flogin1.

**Comment by jasondhales** [comment]
We're still interested in this.

Find log files at falcon1:/projects/bison/git/bison/examples/...
Wed Mar 18 19:38:30 MDT 2015
restart.restart_2......................................................................................... OK [184.667s]
restart.restart_1......................................................................................... OK [187.126s]
restart.restart_3................................................................. FAILED (MISSING GOLD FILE) [185.304s]

**Comment by permcody** [comment]
Yes, I have it 60% complete. Haven't had time to work on it do to either of
the last two sprint problems. I haven't forgotten about it. Thanks for
checking in again.

**Comment by jasondhales** [comment]
Looks like this has fallen by the wayside.  Can we restart it?

**Comment by permcody** [comment]
Yes, I'll find the time to work on this. I had grand plans to refactor the TestHarness but that is a huge project. I think I can get this case working without a whole lot of disturbance to everything else. It won't be quite as elegant but working is better than nothing.

**Comment by permcody** [comment]
This hasn't slipped my mind. I will make sure I assign a resource to this
task at the beginning of the fiscal year.

**Comment by permcody** [comment]
@milljm - I think it's safe to say I'm not going to get back to my rewrite of the whole TestHarness anytime soon or maybe ever. However the recent addition of the DependencyResolver to the TestHarness (https://github.com/idaholab/moose/pull/8595) makes this ticket much more realistic to accomplish with the current design. 

We now have the ability to get back a list of Tester objects with dependencies per directory before we run them. That's exactly what I'm doing in that other PR. We can then use that information to copy test files to the same directory instead of independent directories like we are doing now. This would allow tests to depend on each other even when using the `--pbs` option. You are currently storing a file with launched tests and their directories correct? If so, this is all very doable. Start thinking about it, I'll talk to you more about it next week if you have questions.

**Comment by milljm** [comment]
First, we will need to launch a job and _then_ place that job on hold. There appears to be no way to submit a job inherently in a holding status. Everything I am reading suggests you need to have a JOB ID and use qhold <job id> to place the job on hold after it has been launched.

PBS's example of placing a job on hold:
```bash
[~]$ qsub job.sh
12345.testname.server
[~]$ qhold 12345
```
This will need to be done for each job that requires no dependencies. Thats hurdle 1. Next, we will need to figure out the actual dependencies, and launch those jobs with a special switch that tells it what job id/s it is waiting for:

```bash
#PBS –W depend=afterok:12345:<possible job id of another test>
```
After doing this, we swing back around and release the original 12345 job we put on hold earlier:

```bash
[~]$ qrls 12345
```
We recently discovered an issue with PBS, that if job id 12345 launches and finishes before we launch a job that depends on it, the newly launched dependency job will be stuck in the hold queue forever. Which is why we need to launch the first job and immediately place that job on hold.

Thoughts:
This means for tests that can complete super quick (like less than a second), we might have issues. We _could_ add a sleep command in the actual pbs submission script... but thats hackish!

We can also use the inherently broken functionality of PBS and have that first job wait for an already completed job id, like 1 or something?

**Comment by permcody** [comment]
> We recently discovered an issue with PBS, that if job id 12345 launches and finishes before we launch a job that depends on it, the newly launched dependency job will be stuck in the hold queue forever. Which is why we need to launch the first job and immediately place that job on hold.

Question 1: If we manually call release in this situation. Will the job run then? If so, move to next question:

Question 2: If 1. works, then let's test to see what happens when we have a job with a dependency that is manually released because it's holding on an already completed job, BUT make sure that the dependency hasn't completed yet! If you release the job but it still waits for the dependency then we have a solution to this whole mess.

**Comment by milljm** [comment]
1. The job will run yes.
2. Hmm yeah that might work. 

Just so I am hearing you correctly:

1. We launch Jobs A, B, C... normally.
2. We release _all_ jobs regardless of any job's status because: Jobs B (and C) either continue to hold if Job A (or B,C,D, etc) has not completed (and the system now churns correctly), or the job/s start immediately because of this PBS issues of Job A having completed before B (or C,D,...) launched. Either way, a win?

**Comment by permcody** [comment]
Basically we have only one problem. It's that PBS doesn't do the right thing for jobs on hold if a job has completed before another job begins. I'm hoping that that is the only bug and it can be resolved by always releasing every job regardless of it's status. In fact, if it's already running and we call release, so what. If it's got a dependency that's still running and we call release, so what.

**Comment by permcody** [comment]
@milljm - This should be fixed now. You can test the example provided in the description to make sure. It would be awesome to close this very old ticket.

**Comment by milljm** [comment]
@jasondhales @permcody Yes this has been solved when we merged #10046. Those restart tests were the actual tests I was using in order to make sure things did what they were supposed to do (launch in the right order of 1, 2, 3 and test in the right order; 3, 2, 1).

Ran it again this morning:

```pre
[milljm@rod][/ssd/milljm/bison]> ./run_tests -i examples --pbs jason_tests --re=restart
restart.restart_1............................................................................ LAUNCHED 754789
restart.restart_2............................................................................ LAUNCHED 754790
restart.restart_3............................................................................ LAUNCHED 754791
-------------------------------------------------------------------------------------------------------------
Your session file is jason_tests
Ran 0 tests in 0.8 seconds
0 passed, 0 skipped, 3 pending, 0 failed
[milljm@rod][/ssd/milljm/bison]> ./run_tests -i examples --pbs jason_tests --re=restart
restart.restart_3.................................................................................... HOLDING
restart.restart_2.................................................................................... RUNNING
restart.restart_1......................................................................................... OK
-------------------------------------------------------------------------------------------------------------
Your session file is jason_tests
Ran 0 tests in 0.3 seconds
1 passed, 0 skipped, 2 pending, 0 failed
[milljm@rod][/ssd/milljm/bison]> ./run_tests -i examples --pbs jason_tests --re=restart
restart.restart_3......................................................................................... OK
restart.restart_2......................................................................................... OK
restart.restart_1......................................................................................... OK
-------------------------------------------------------------------------------------------------------------
Your session file is jason_tests
Ran 3 tests in 0.9 seconds
3 passed, 0 skipped, 0 pending, 0 failed
```

**Comment by permcody** [comment]
Closed #10046

**Comment by milljm** [comment]
🎉
